lib/layer.js

```diff
@@ -54,11 +54,22 @@ function Layer(p, options, fn) {
   this.name = fn.name || '<anonymous>'
   this.params = undefined
   this.path = undefined
-  this.regexp = pathRegexp(path, this.keys = [], opts)
+  this.matchedPath = undefined
 
   // set fast path flags
-  this.regexp.fast_star = path === '*'
-  this.regexp.fast_slash = path === '/' && opts.end === false
+  this.fastStar = paths === '*'
+  this.fastSlash = paths === '/' && opts.end === false
+
+  this.paths = (!Array.isArray(paths) ? [paths] : paths).map(function (path) {
```

Comment:
So I assume that the purpose of pre-expanding the array of paths is to store data long with them? If that is the case `path-to-regexp` supports arbitrarily-nested arrays like `["/foo",["/bar","/fizz"]]` and this implementation would end up only accounting for the first level.

Reply:
One of the big changes in this PR is to always call `path-to-regexp` with a single path segment instead of an array of paths. That allows us to identify the matching path element in the array of paths (it's the one that returns a match from the regex). But as you point out, that won't work for deeply nested arrays. 

Perhaps a better approach is to use the return value of the regex to indicate which element in the array was the match. I'm not totally sure what that would look like for the deeply nested array case, or if that's feasible without making breaking changes to `path-to-regexp`, but it would allow us to avoid pre-expansion of the array here.

Reply:
Is `["/foo",["/bar","/fizz"]]` equivalent to `["/foo", "/bar", "/fizz"]`? If it is, then recursively flattening the array of paths should be fine.

However, another issue is if evaluating multiple regexes hurts performance compared to evaluating a single regex with clauses separated by `|`. Is that something to worry about?

Reply:
> Is ["/foo",["/bar","/fizz"]] equivalent to ["/foo", "/bar", "/fizz"]? If it is, then recursively flattening the array of paths should be fine.

Currently this module delegates the argument to `path-to-regexp` and I don't know the answer off-hand and would have to look. If you're not even sure, then I think the entire way this functions needs to be re-examined, as if the code here is making assumptions for how `path-to-regexp` works, we should have a good grasp on how it is working.

>However, another issue is if evaluating multiple regexes hurts performance compared to evaluating a single regex with clauses separated by |. Is that something to worry about?

It would probably depending on how much of a performance penalty it has. Typically adding features has a performance impact, but I it is really large, we may want to gate the feature. It would depend on (1) what % of users are going to use the feature and (2) how big of a penalty it has. For example, if 99% of users will use it, then it probably doesn't matter. But if only like 10% of users will use it and it will hurt performance by 20%, maybe a gate would be in order.

Reply:
> Currently this module delegates the argument to path-to-regexp and I don't know the answer off-hand and would have to look. If you're not even sure, then I think the entire way this functions needs to be re-examined, as if the code here is making assumptions for how path-to-regexp works, we should have a good grasp on how it is working.

A quick test of path-to-regexp indicates they are functionally equivalent (the only difference being an extra non-capturing group wrapper for each level of nesting):
```js
pathToRegexp(['/', '/']) // /(?:^\/(?:\/)?$|^\/(?:\/)?$)/i
pathToRegexp(['/', ['/']]) // /(?:^\/(?:\/)?$|(?:^\/(?:\/)?$))/i
```

Looking at [the source](https://github.com/pillarjs/path-to-regexp/blob/7b4078db2cfc929eb287432be0ecd3798b0de820/src/index.ts#L477-L484), `path-to-regexp` does appear to recursively expand the array, so we should be able to do the same before calling it. 

> But if only like 10% of users will use it and it will hurt performance by 20%, maybe a gate would be in order.

Any theoretical hit to performance should only affect the use of arrays. So that should help.

Reply:
> Looking at the source, path-to-regexp does appear to recursively expand the array, so we should be able to do the same before calling it.

Be carful to look at the proper version of the module :) Your link is about version 6.1.0 module, several major versions ahead of what his module is actually using.

Reply:
Good catch, although looks like the [implementation hasn't changed since v3.0.0](https://github.com/pillarjs/path-to-regexp/blob/796f3fdae1186f3ef7afe029555e2b141ab9ece7/index.js#L263-L270).

I've updated the PR to call `array-flatten` on the paths.